### ISSUE TITLE
perf(direct): batch _process_batch GPU uploads (eliminate per-frame upload loop)

### ISSUE THREAD

**Issue by chunzhe10** [chunzhe10/dorea#68]
## Problem

After fp16 + InstanceNorm fix landed (chunzhe10/dorea#67), direct mode is at 4.25 fps. The next bottleneck is `_process_batch()` in \`python/dorea_inference/raune_filter.py\`:

\`\`\`python
def _process_batch(batch_frames_np, model, normalize, fw, fh, pw, ph, transfer_fn, model_dtype):
    proxy_tensors = []
    for rgb_np in batch_frames_np:
        # Per-frame upload — N serialized H2D transfers
        full_t = torch.from_numpy(rgb_np).cuda().float() / 255.0
        full_t = full_t.permute(2, 0, 1).unsqueeze(0)
        proxy_t = F.interpolate(full_t, size=(ph, pw), ...)
        proxy_norm = (proxy_t - 0.5) / 0.5
        proxy_tensors.append(proxy_norm.squeeze(0))
        del full_t  # full-res tensor freed before transfer step!

    proxy_batch = torch.stack(proxy_tensors).cuda()  # double .cuda() (no-op on already-cuda tensors)
    ...

    # Apply transfer per frame — re-uploads the same numpy buffer
    for i in range(n):
        full_t = torch.from_numpy(batch_frames_np[i]).cuda().float() / 255.0  # SECOND UPLOAD
        ...
\`\`\`

## Impact

Two performance reviewers (PR #65 and PR #67) flagged this. At 4K with batch=8:
- ~50MB wasted PCIe traffic per frame (full-res frame uploaded twice)
- Per-frame loop serializes N CUDA stream sync points (vs 1 batched upload)
- batch=8 gives only +0.7% throughput vs batch=4 because the per-frame loop scales linearly
- batch=16 actively REGRESSES (3.47 fps) due to compounding overhead

## Proposed fix

Replace both per-frame loops with batched operations:

\`\`\`python
def _process_batch(batch_frames_np, model, normalize, fw, fh, pw, ph, transfer_fn, model_dtype):
    n = len(batch_frames_np)

    with torch.no_grad():
        # Single batched H2D transfer
        rgb_np_stack = np.stack(batch_frames_np)  # (N, H, W, 3) uint8
        full_batch = torch.from_numpy(rgb_np_stack).cuda().float() / 255.0
        full_batch = full_batch.permute(0, 3, 1, 2)  # (N, 3, H, W)

        # Batched downscale to proxy
        proxy_batch = F.interpolate(full_batch, size=(ph, pw), mode="bilinear", align_corners=False)
        proxy_norm = (proxy_batch - 0.5) / 0.5

        # Cast to model dtype
        if proxy_norm.dtype != model_dtype:
            proxy_norm = proxy_norm.to(model_dtype)

        # RAUNE inference
        raune_out = model(proxy_norm).float()
        # ... (existing OKLab delta path)

        # Apply transfer batched (full_batch already on GPU — reuse it!)
        result_batch = transfer_fn(full_batch, delta_full)  # batched transfer

        # Single batched D2H
        results_u8 = (result_batch.permute(0, 2, 3, 1).clamp(0, 1) * 255.0).to(torch.uint8).cpu().numpy()
        return [results_u8[i] for i in range(n)]
\`\`\`

## Expected outcome

- Larger batch sizes (8, 16) actually deliver throughput improvements
- batch=8 should reach ~5-5.5 fps (target)
- batch=16 should not regress
- Cumulative target: 4.25 → ~5.5-6 fps

## Constraints

- transfer_fn (Triton kernel and PyTorch fallback) must accept batched input — verify before refactoring
- Output must remain bit-identical to current per-frame loop
- Quality preserving only

## Reference

- Performance reviewer flagged in PR #65 (issue #64) and PR #67 (issue #66)
- Bottleneck analysis: corvia entries 019d74c9, 019d7501
- This is part of the GPU optimization sweep #66 — should be PR 1.5/3 (between fp16 and torch.compile)

**Comment by chunzhe10** [comment]
🤖 Claimed by Claude Code agent — starting dev-loop.

**Comment by chunzhe10** [comment]
## Closing as wontfix — plan premise was incorrect

Implemented the batched upload refactor on branch `feat/68-batched-uploads` and measured results. The plan's core premise — that per-frame uploads in `_process_batch()` were the next bottleneck — did not hold up empirically.

### Measured results

| Config | Before (per-frame loop) | After (batched) |
|---|---|---|
| batch=4 | 4.25 fps | 4.17 fps (neutral, within noise) |
| batch=8 | 4.36 fps | 4.23 fps (neutral, within noise) |
| batch=16 | 3.47 fps | **STALLS** — VRAM exhausted at 5141/6144 MB |

- `gpu_busy` per-frame: 230 → 232 ms (unchanged)
- Visual diff: YAVG=0.117 vs bit-exact baseline (rounding order change from reshape/permute)

### Why the optimization didn't work

My mental model was wrong. Per-frame H2D uploads were NOT serializing the GPU — they were happening **concurrently** with RAUNE forward-pass compute via CUDA streams. The "N serial PCIe transfers" I was worried about were already hidden behind the 230ms RAUNE forward pass. Eliminating them saved nothing measurable.

**The actual bottleneck is RAUNE compute itself** (232 ms/frame at batch=8 fp16 on RTX 3060), not the upload loop. Fixing upload overhead on a GPU-compute-bound workload is a no-op.

### Why batch=16 regressed to a stall

The "reuse full_batch" approach held the entire full-resolution tensor on GPU through the whole pipeline. At batch=16 with 4K fp32:
- `full_batch`: 16 × 3 × 2160 × 3840 × 4 bytes = 1.6 GB
- `delta_full` fp32: 1.6 GB
- Triton wrapper permute+contiguous fp16 copies: ~800 MB
- RAUNE activations + model weights: ~1 GB
- Output buffer fp16: ~800 MB
- **Total ≈ 5.8 GB on a 6 GB card → allocator hangs in fragmentation territory**

The old per-frame path only ever held 1 frame's `full_t` transient in the transfer loop, so batch=16 had ~1.6 GB more VRAM headroom.

### Lesson learned

Before proposing upload-batching optimizations, verify that the uploads are actually on the critical path. On a GPU-compute-bound workload with CUDA streams, H2D transfers can overlap with compute and don't need to be minimized.

### Next steps

Closing this issue. The real lever for direct mode throughput is RAUNE compute itself — proceeding with issue #66 PR 2/3 (`torch.compile(model.half())`) and eventually PR 3/3 (NVDEC). Those attack the actual bottleneck.

Finding recorded in corvia for future reference.
